### PR TITLE
[JS Api] Reuse C-Api for emitText and emitStackIR

### PR DESCRIPTION
**Pull request by MaxGraey**
Also fixed old bug which caused new lines to be displayed incorrectly

### CODE REVIEW

**Comment by MaxGraey** [comment]
@tlively You know C++ much better than me. I can't understand why all new lines appended into beginning by ostringstream / stringstream?
```diff
+
+
+
+
+
+
+
+
 (module
  (type $i (func (param i32) (result i32)))
  (memory $0 0)
  (export "test" (func $test))
  (func $test (param $0 i32) (result i32)
-  block $block0 (result i32)
-   local.get $0
-   if (result i32)
-    local.get $0
-   else
-    i32.const 0
-   end
-  end
- )
+  block $block0 (result i32)   local.get $0   if (result i32)    local.get $0   else    i32.const 0   end  end )
 )
```

**Comment by MaxGraey** [comment]
@kripken @tlively could you take a look?

**Comment by MaxGraey** [comment]
new approach for `doIndent` should decrease code size. See godbolt:
https://godbolt.org/z/KMYMdn7z5

**Review by tlively — COMMENTED** [review]
It looks like a lot of this diff doesn't change behavior. Can you make this diff cover just the behavior changes?

**Comment by MaxGraey** [comment]
> Can you make this diff cover just the behavior changes?

In general, there shouldn't have been any changes. It's more refactoring. The only thing was a slight modification of the C-API (this is added to CHANGELOG.md) and [one bug was fixed](https://github.com/WebAssembly/binaryen/pull/4832/files#diff-a5594d8c2551220856242429b27daca74d0eca899ffe8486b1f2806c3f862084R3626). These changes allow you to sync C-Api such as `BinaryenModuleAllocateAndWriteStackIR` and `BinaryenModuleAllocateAndWriteText` with the JS-API `emitStackIR` and `emitText` and minimize the work in the glue code itself.

**Review by tlively — COMMENTED** [review]
I'm not sure the cosmetic changes are worth it. Changing stringstream to ostringstream or strncopy to std::copy_n seems like unnecessary churn. I'm also not sure it makes sense to have an `optimize` option on API functions that otherwise just print stuff. The benefit in the JS glue code seems small.

**Comment by MaxGraey** [comment]
> The benefit in the JS glue code seems small.

These changes align behavior for JS-APi and C-Api. For example, JS-Api has the "optimize" parameter, while methods from C-Api do not. Also, the methods in JS-Api use a different set of calls, especially this is noticeable for `emitStackIR`. Having reused and unified the API, all further changes will only be made in the C-API, which is very convenient. Also, JS-API used workaround by hooking output stream and concatenate string parts in closure which looked rather inefficient.

**Comment by MaxGraey** [comment]
About std::copy_n instead of strncpy. `std::copy_n` will be more efficient since it will be lowered to memove. By the way, it is `std::copy_n` that is [used for sourcemaps](https://github.com/WebAssembly/binaryen/blob/main/src/binaryen-c.cpp#L4172). I decided to unify this in the same way.

File: src/passes/Print.cpp [hunk 1]
```diff
@@ -3627,15 +3623,18 @@ printStackIR(StackIR* ir, std::ostream& o, Function* func) {
...
   return o;
 }
 
-std::ostream& printStackIR(std::ostream& o, Module* module) {
+std::ostream& printStackIR(std::ostream& o, Module* module, bool optimize) {
   wasm::PassRunner runner(module);
   runner.add("generate-stack-ir");
+  if (optimize) {
+    runner.add("optimize-stack-ir");
+  }
```
Comment (dcodeIO) [inline src/passes/Print.cpp:3637]: Regarding having `optimize` on API functions, perhaps here is why it makes sense: `printStackIR` already runs a pass (here: `generate-stack-ir`), so not just printing, and putting the optimize pass directly behind it seems good to do? The behavior is then made controllable via the C API functions calling `printStackIR`, with the side-effect of also matching what the JS API provides. On the AS side, we are always setting `optimize=true` to remove sequences of `unreachable`s and similar iirc.

File: src/dataflow/utils.h [hunk 2]
```diff
@@ -32,11 +32,7 @@
 namespace wasm::DataFlow {
 
 inline std::ostream& dump(Node* node, std::ostream& o, size_t indent = 0) {
-  auto doIndent = [&]() {
-    for (size_t i = 0; i < indent; i++) {
-      o << ' ';
-    }
-  };
+  auto doIndent = [&]() { o << std::string(indent, ' '); };
```
Comment (tlively) [inline src/dataflow/utils.h:35]: I like this change, but is it connected to any other changes in this PR? If not, can we separate it into a separate PR?

**Review by tlively — APPROVED** [review]

File: src/dataflow/utils.h (hunk 2 above)
Comment (MaxGraey) [reply src/dataflow/utils.h:35]: Sure! Will separate to another PR
